File: VBNetCore.Mvvm/Helpers/UiExecution.vb
Imports VBNetCore.Mvvm.Abstractions
Imports System
Imports System.Threading.Tasks
Imports System.Windows
Imports System.Windows.Threading

Namespace VBNetCore.Mvvm.Helpers
    Public Class UiExecution
        Implements IUiExecution
        Public Sub Execute(ByVal action As Action, ByVal Optional priority As DispatcherPriority = DispatcherPriority.DataBind) Implements IUiExecution.Execute
            Dim dispatcher = If(Application.Current.Dispatcher, Threading.Dispatcher.CurrentDispatcher)
            dispatcher.Invoke(action, priority)
        End Sub

        Public Function ExecuteAsync(ByVal action As Action, ByVal Optional priority As DispatcherPriority = DispatcherPriority.DataBind) As Task Implements IUiExecution.ExecuteAsync
            Dim dispatcher = If(Application.Current.Dispatcher, Threading.Dispatcher.CurrentDispatcher)
            Return dispatcher.InvokeAsync(action, priority).Task
        End Function
    End Class
End Namespace

File: VBNetCore.Mvvm/Helpers/UiExecution.cs
using NetCore.Mvvm.Abstractions;
using System;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Threading;

namespace NetCore.Mvvm.Helpers
{
    public class UiExecution : IUiExecution
    {
        public void Execute(Action action, DispatcherPriority priority = DispatcherPriority.DataBind)
        {
            var dispatcher = Application.Current.Dispatcher ?? Dispatcher.CurrentDispatcher;
            dispatcher.Invoke(action, priority);
        }

        public Task ExecuteAsync(Action action, DispatcherPriority priority = DispatcherPriority.DataBind)
        {
            var dispatcher = Application.Current.Dispatcher ?? Dispatcher.CurrentDispatcher;
            return dispatcher.InvokeAsync(action, priority).Task;
        }
    }
}
